### ISSUE TITLE
Proposed Analyzer: Prevent invocation of non-pure methods from readonly value-type fields

### ISSUE THREAD

**Issue by maxkoshevoi** [dotnet/roslyn-analyzers#3523]
This issue describes analyzer added by #3520

When value type field declared as `readonly`, you cannot changes it's state outside of constructor.
And if you try to directly change it, you get expected compile error:
![image](https://user-images.githubusercontent.com/6609929/79750211-3c75de00-8319-11ea-9ebe-60d56393f7db.png)

But, if you try to call non-pure method from within said value type, you won't get compile- nor runtime error, but changes made by this method will be lost:
![image](https://user-images.githubusercontent.com/6609929/79750511-c32abb00-8319-11ea-8b9a-c5235cbc3e53.png)

This can cause really unobvious behavior. For example:
```cs
class ListWrapper
{
    private readonly List<int>.Enumerator _enumerator;
 
    public ListWrapper(List<int> list)
    {
        _enumerator = list.GetEnumerator();
    }
 
    public void PrintNextElement()
    {
        _enumerator.MoveNext();
        Console.Write(_enumerator.Current);
    }
}

var wrapper = new ListWrapper(new List<int> { 1, 2, 3 });
wrapper.PrintNextElement();
wrapper.PrintNextElement();
wrapper.PrintNextElement();
```
Output will be: 000
But expected output is: 123

And as I said, there will be no warnings or errors on compile and runtime.
Here is even post about it from jetbrains: https://www.jetbrains.com/help/resharper/ImpureMethodCallOnReadonlyValueField.html

**Comment by mavasani** [comment]
@sharwell for approval/discussion
